### PR TITLE
Issue #7867: Move EngineSession from SessionManager to BrowserState.

### PR DESCRIPTION
**Pull request by pocmo**
Fenix PR: https://github.com/mozilla-mobile/fenix/pull/13933
R-B PR: https://github.com/mozilla-mobile/reference-browser/pull/1292

---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by pocmo** [comment]
bors try

**Comment by pocmo** [comment]
> Uh oh! Looks like an error! Details
> Failed to get your artifact.

I think that should be fixed after this lands:
https://github.com/mozilla-mobile/android-components/pull/8077

**Comment by csadilek** [comment]
Fixed tests for `feature-intent`, `feature-customtabs`, `feature-contextmenu`, `browser-session`. Almost through :).

**Comment by csadilek** [comment]
Fixed tests for `feature-search` and `feature-tabs`.

**Comment by csadilek** [comment]
Found a leak when testing, but it's unrelated to our change here: https://github.com/mozilla-mobile/fenix/issues/13785

**Comment by csadilek** [comment]
`feature-session` tests needed some work. Should be green now.

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/EngineState.kt [hunk 1]
```diff
@@ -13,8 +13,10 @@ import mozilla.components.concept.engine.EngineSessionState
  * @property engineSession the engine's representation of this session.
  * @property engineSessionState serializable and restorable state of an engine session, see
  * [EngineSession.saveState] and [EngineSession.restoreState].
+ * @property engineObserver TODO
```
Comment (csadilek) [inline components/browser/state/src/main/java/mozilla/components/browser/state/state/EngineState.kt:16]: Let's add a docs here.

File: components/feature/p2p/src/main/java/mozilla/components/feature/p2p/P2PFeature.kt [hunk 2]
```diff
@@ -132,9 +132,16 @@ class P2PFeature(
             return
         }
 
-        val engineSession = sessionManager.getOrCreateEngineSession(session)
-        val messageHandler = P2PContentMessageHandler()
-        extensionController?.registerContentMessageHandler(engineSession, messageHandler)
+        // TODO Action for registering content message handler? Do we really need that?
```
Comment (csadilek) [inline components/feature/p2p/src/main/java/mozilla/components/feature/p2p/P2PFeature.kt:135]: Let's look into this quickly before landing and file a follow-up if we can't address now.

File: components/feature/p2p/src/main/java/mozilla/components/feature/p2p/P2PFeature.kt [hunk 3]
```diff
@@ -160,12 +167,19 @@ class P2PFeature(
...
-            activeSession?.let {
-                extensionController?.sendContentMessage(
-                    json,
-                    sessionManager.getOrCreateEngineSession(it)
-                )
-            }
+            val session = activeSession ?: return
+
+            // TODO Action for sending content message. Do we really need that?
+            json.hashCode()
```
Comment (csadilek) [inline components/feature/p2p/src/main/java/mozilla/components/feature/p2p/P2PFeature.kt:173]: 🤔 looks like left-overs we can remove?

File: components/browser/state/src/main/java/mozilla/components/browser/state/state/EngineState.kt (hunk 1 above)
Comment (csadilek) [reply components/browser/state/src/main/java/mozilla/components/browser/state/state/EngineState.kt]: Fixed.

File: components/feature/p2p/src/main/java/mozilla/components/feature/p2p/P2PFeature.kt (hunk 2 above)
Comment (csadilek) [reply components/feature/p2p/src/main/java/mozilla/components/feature/p2p/P2PFeature.kt]: Updated the `P2PFeature` so it doesn't use `SessionManager` anymore and fixed the TODOs.

File: components/feature/p2p/src/main/java/mozilla/components/feature/p2p/P2PFeature.kt (hunk 3 above)
Comment (csadilek) [reply components/feature/p2p/src/main/java/mozilla/components/feature/p2p/P2PFeature.kt]: Fixed.

**Comment by csadilek** [comment]
Fenix PR: https://github.com/mozilla-mobile/fenix/pull/13933
R-B PR: https://github.com/mozilla-mobile/reference-browser/pull/1292

File: components/lib/state/src/main/java/mozilla/components/lib/state/internal/StoreThreadFactory.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * Custom [ThreadFactory] implementation wrapping [Executors.defaultThreadFactory] that allows
+ * asserting whether a caller is on the created thread.
+ *
+ * For usage with [Executors.newSingleThreadExecutor]: Only the last created thread is kept and
+ * compared when [assertOnThread] is called.
+ */
+internal class StoreThreadFactory : ThreadFactory {
+    private var thread: Thread? = null
```
Comment (csadilek) [inline components/lib/state/src/main/java/mozilla/components/lib/state/internal/StoreThreadFactory.kt:18]: @pocmo looks like we need this to be `volatile`.

**Pull request merged**